cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        synchronized (lastNotificationLock) {
+          lastNotification = new StreamSizeNotification(baseTs, baseSize);
+        }
+      } else {
+        newTask.startSchedule(baseRunSize, baseRunTs, active, persist);
+      }
+
+      if (active) {
+        activeTasks.incrementAndGet();
+      }
+
+      if (lastNotification != null) {
+        // In all cases, when creating a schedule - either if it comes from the store or not,
+        // we want to pass it the last seen notification
+        // This call will send the notification to all the active tasks. The ones which are active before
+        // that method is called will therefore see the notification twice. It is fine though.
+        received(lastNotification, null);
+      }
+    }
+
+    /**
+     * Suspend a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void suspendScheduleTask(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null) {
+        return;
+      }
+      synchronized (this) {
+        if (task.suspend()) {
+          activeTasks.decrementAndGet();
+        }
+      }
+    }
+
+    /**
+     * Resume a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void resumeScheduleTask(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null || !task.resume()) {
+        return;
+      }
+      if (activeTasks.incrementAndGet() == 1) {
+        // There were no active tasks until then, that means polling the stream was disabled.
+        // We need to check if it is necessary to poll the stream at this time, if the last
+        // notification received was too long ago, or if there is no last seen notification
+        synchronized (lastNotificationLock) {
+          if (lastNotification == null ||
+            (lastNotification.getTimestamp()
+              + TimeUnit.SECONDS.toMillis(pollingDelay) <= System.currentTimeMillis())) {
+            long streamSize = pollStream();
+            lastNotification = new StreamSizeNotification(System.currentTimeMillis(), streamSize);
+          }
+        }
+      }
+      task.received(lastNotification);
+    }
+
+    /**
+     * Delete a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void deleteSchedule(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask scheduleTask = scheduleTasks.remove(getScheduleId(programId, programType, scheduleName));
+      if (scheduleTask != null && scheduleTask.isRunning()) {
```

Comment:
you may have to be careful. Suppose a task is suspended, but concurrently being resumed. Then the resuming thread will not know that this thread is deleting. It will hence increment the activeTasks by one, but the task is deleted. The consequence is (I believe) that we will never stop polling.